src/ctapipe/monitoring/interpolation.py

```diff
@@ -186,3 +183,108 @@ def add_table(self, tel_id, input_table):
+        """
+
+        self._check_interpolators(tel_id)
+
+        if isinstance(columns, str):
+            columns = [columns]
+
+        result = {}
+        mjd = time.to_value("mjd")
+        for column in columns:
+            if column not in self._interpolators[tel_id]:
+                raise ValueError(
+                    f"Column '{column}' not found in interpolators for tel_id {tel_id}"
+                )
+            result[column] = self._interpolators[tel_id][column](mjd)
+
+        if len(result) == 1:
+            return result[columns[0]]
+        return result
+
+    def add_table(self, tel_id: int, input_table: Table, columns: list[str]) -> None:
+        """
+        Add a table to this interpolator for specific columns.
+
+        Parameters
+        ----------
+        tel_id : int
+            Telescope id.
+        input_table : astropy.table.Table
+            Table of values to be interpolated, expected columns
+            are ``start_time`` as ``validity start Time`` column,
+            ``end_time`` as ``validity end Time`` and the specified columns
+            for the data of the chunks.
+        columns : list of str
+            Names of the columns to interpolate.
+        """
+
+        required_columns = set(self.required_columns)
+        required_columns.update(columns)
+        self.required_columns = frozenset(required_columns)
+        self._check_tables(input_table)
+
+        input_table = input_table.copy()
+        input_table.sort("start_time")
+        start_time = input_table["start_time"].to_value("mjd")
+        end_time = input_table["end_time"].to_value("mjd")
+
+        if tel_id not in self._interpolators:
+            self._interpolators[tel_id] = {}
+
+        for column in columns:
+            values = input_table[column]
+
+            def interpolate_chunk(
+                mjd: float, start_time=start_time, end_time=end_time, values=values
+            ) -> float:
+                # Find the index of the closest preceding start time
+                preceding_index = np.searchsorted(start_time, mjd, side="right") - 1
+                if preceding_index < 0:
+                    return np.nan
+
+                # Check if the time is within the valid range of the chunk
+                if start_time[preceding_index] <= mjd <= end_time[preceding_index]:
+                    value = values[preceding_index]
+                    if not np.isnan(value):
+                        return value
+
+                # If the closest preceding chunk has nan, check the next closest chunk
+                for i in range(preceding_index - 1, -1, -1):
+                    if start_time[i] <= mjd <= end_time[i]:
+                        value = values[i]
+                        if not np.isnan(value):
+                            return value
+
+                return np.nan
+
+            self._interpolators[tel_id][column] = interpolate_chunk
```

Comment:
I think it's a bit weird here to store closures in a dict instead of just calling a method with `tel_id` and `column` that then checks if the necessary table exists.

Reply:
ok, i'll change it to a method

Reply:
The current implementation follows the PointingInterpolator approach, but of course can be changed.

Reply:
so, ``_check_interpolators`` in the parent class assumes the current implementation. If i change it to what you describe i will need to make changes to the parent class or forgo this check for the ``ChunkInterpolator``

Reply:
Maybe keeping it like this is better.

Reply:
Ok, i can make this as a separate class.

The way overlapping chunks are handled is that it tries to return the chunk with the earlies start time and use later chunks if that chunk has no valid data data (for instance there was a car flash during the period that causes the calibration data to be flagged as invalid). This way the amount of time for which valid data is provided is maximized.

Reply:
Also i want to add that my intent was to use it in a similar way to the ``PointingInterpolator``. When event data are read from some file with the ``EventSource`` or some similar tool then the ``PointingInterpolator`` provides event-by-event data for the pointing. I would like this interpolator to do the same for calibration data. The Calibrators can then just look up event-by-event calibration data provided by this class and apply it to the events.

Reply:
I consider a _new_ data point as a point in (value, timestamp) domain :) Prior to the call, there's no unambiguous pair value-timestamp, so to me it is new.

I agree that it is a bit weird, however, this logically fits in the inheritance tree and in the future we may want to introduce a more complex interpolation (i.e. linear between the nearest measured points) and I would like to preserve the API.

Reply:
@maxnoe could i suggest that i open an issue for refactoring the pointing interpolator to use a method as you describe. I can implement this change then on this new interpolator and the PointingInterpolator after this request is merged. Switching both over like that would not impact the API.

Reply:
Wait: what is wrong with the pointing interpolator now?

Reply:
It also keeps the call results in a mutable shared state (interpolators dict).

Reply:
> the call results

Not the call results. The interpolator instances.

That was not the point... the point was using a closure to store data on a class, instead of the actual datastructure that can be e.g. investigated in a test or debugger.

Reply:
Is the point that the `Interpolator` baseclass already assumes this? That's the case because it was designed as the base class to share code for things using `interp1d` to interpolate tables.

You have something completely different here now. So it should not use the same base class. 

If we need a common base class to ensure a common API for anything that takes some monitoring tables and returns the correct value for a given time, we should have a purely abstract interface definition and then the current `Interpolator` only as  a baseclass for the things that are based on `interp1d`.

So:

```
class MonitoringInterpolator:
   @abstractmethod 
   def __call__(self, timestamp):
        ...

class LinearMonitoringInterpolator(MonitoringInterpolator):
     ...

class PointingInterpolator(LinearMonitoringInterpolator):
   ...

class TimeRangeFinder(MonitoringInterpolator)
   ...
```

Reply:
@maxnoe So, i can move the closure and make it just a method of ChunkInterpolator. If i do that i don't need to do anything that impacts the API. Would that be okay?

Reply:
The `Interpolator` class has traitlets that are directly related to `interp1d` for example. That really doesn't apply here, so I don't think it should be in a shared baseclass.

Reply:
ok. @ctoennis please provide an interface sketch, maintaining one abstract interface-only class, so we can check it before the implementation.

Reply:
ok, I'm doing that

Reply:
Here is a sketch for the parent classes:

``
class MonitoringInterpolator(Component, metaclass=ABCMeta):
    """
    MonitoringInterpolator parent class.

    Parameters
    ----------
    h5file : None | tables.File
        An open hdf5 file with read access.
    """

    telescope_data_group = None
    required_columns = set()
    expected_units = {}

    def __init__(self, h5file: None | tables.File = None, **kwargs: Any) -> None:
        super().__init__(**kwargs)
 
        ...

    @abstractmethod
    def add_table(self, tel_id: int, input_table: Table) -> None:
        """
        Parameters
        ----------
        tel_id : int
            Telescope id.
        input_table : astropy.table.Table
            Table of pointing values, expected columns
            are always ``time`` as ``Time`` column and
            other columns for the data that is to be interpolated.
        """
        pass

    def _check_tables(self, input_table: Table) -> None:
        ...

    def _check_interpolators(self, tel_id: int) -> None:
        ...

    def _read_parameter_table(self, tel_id: int) -> None:
        ...

class LinearInterpolator(MonitoringInterpolator):
    """
    LinearInterpolator parent class.

    Parameters
    ----------
    h5file : None | tables.File
        An open hdf5 file with read access.
    """

    bounds_error = traits.Bool(
        default_value=True,
        help="If true, raises an exception when trying to extrapolate out of the given table",
    ).tag(config=True)

    extrapolate = traits.Bool(
        help="If bounds_error is False, this flag will specify whether values outside"
        "the available values are filled with nan (False) or extrapolated (True).",
        default_value=False,
    ).tag(config=True)

    def __init__(self, h5file: None | tables.File = None, **kwargs: Any) -> None:
        super().__init__(**kwargs)

        self.interp_options: dict[str, Any] = dict(assume_sorted=True, copy=False)
        if self.bounds_error:
            self.interp_options["bounds_error"] = True
        elif self.extrapolate:
            self.interp_options["bounds_error"] = False
            self.interp_options["fill_value"] = "extrapolate"
        else:
            self.interp_options["bounds_error"] = False
            self.interp_options["fill_value"] = np.nan

        self._interpolators = {}

``

Reply:
:+1:  Looks good

I am not sure if putting the private methods you listed here to the base class will work out in the end, but you'll see how much code can be shared between the chunk interpolator and the linear interpolator.

Reply:
Shall we take an opportunity and stop using `interp1d`? It is legacy, and`scipy` suggests this: https://docs.scipy.org/doc/scipy/tutorial/interpolate/1D.html#tutorial-interpolate-1dsection

Also, if we call a base class LinearInterpolator, I guess we should enforce linearity (in the example above, e.g. cubic spline can be applied)

Reply:
There is no replacement for linear 1d interpolation for interp1d. It is legacy, but not deprecated.

See https://github.com/scipy/scipy/issues/21258